### PR TITLE
Independent constraint

### LINKED ISSUE #50496: distributions.Independent does not correctly update distribution's support
## 🐛 Bug

When a batch dimension is reinterpreted to be an event dimension, the number of event dimensions should increase by 1 (e.g. a distribution over scalars becomes a distribution over vectors) and as a result the distribution's support should be updated. 

Because the distribution's support is not updated, transforms inferred by `biject_to` are applied elementwise (rather than summing/squeezing out the `event_dim`) and methods like `log_abs_det_jacobian` do not correctly compute the determinant of what now should be a multivariate transform (here, this multivariate log determinant is the sum of the elementwise log determinants).

Maybe we should call it `log_abs_diag_jacobian` here instead...? Just kidding.

## To Reproduce

When I run
```
import torch.distributions as dist

dist.Independent(dist.Normal(torch.zeros(1), torch.ones(1)), 1).support == dist.Normal(torch.zeros(1), torch.ones(1)).support
```
I find that the supports are identical before and after reinterpreting batch dimensions.

As a result, the following snippet:
```
import torch.distributions as dist
from torch.distributions.constraint_registry import biject_to

broken_dist = dist.Independent(dist.Exponential(torch.ones(2)), 1)
tform = biject_to(broken_dist.support)

x = torch.randn(3,2)
y = tform(x)  # NOTE: this is applied element-wise across x
tform.log_abs_det_jacobian(x, y).shape. # because broken_dist.event_shape == [2], after taking determinant this should be [3]
```
should report `torch.Size([3])` because taking the log determinant removes the `event_dim`, but instead it reports `[3,2]`.

## Expected behavior

After reinterpreting batch dimensions, the left-hand side is now a distribution over 1-dimensional rank-one tensors. Therefore, I expect the support to be `constraints.real_vector` in this example. More generally, I expect the support of a `dist.Independent` to be a product over the reinterpreted batch dimension of the individual supports, and for the inferred `transforms` to properly sum over the new `event_dim`.

## Environment

```
Collecting environment information...
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.19.1

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] botorch==0.3.3
[pip3] gpytorch==1.3.0
[pip3] numpy==1.19.4
[pip3] numpyro==0.4.1
[pip3] torch==1.7.0
[pip3] torchvision==0.8.1
[conda] Could not collect
```

## Additional context

cc @fritzo @neerajprad @alicanb @vishwakftw @nikitaved

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #50496 

This fixes a number of inconsistencies in torch.distributions.constraints as used for parameters and supports of probability distributions.
- Adds a `constraints.independent` and replaces `real_vector` with `independent(real, 1)`. (this pattern has long been used in Pyro)
- Adds an `.event_dim` attribute to all constraints.
- Tests that `constraint.check(data)` has the correct shape. (Previously the shapes were incorrect).
- Adds machinery to set static `.is_discrete` and `.event_dim` for `constraints.dependent`.
- Fixes constraints for a number of distributions.

## Tested
- added a new check to the constraints tests
- added a new check for `.event_dim`

cc @fehiepsi @feynmanliang @stefanwebb

### CODE REVIEW

**Comment by feynmanliang** [comment]
Fixes #50496

**Review by fehiepsi — APPROVED** [review]
I really like the proposals:
+ adding `independent` constraint to beautifully correct the constraints/supports of Independent and other parameters (such as those in LowRank distribution)
+ having `is_discrete` attribute to those supports. This would simplify users' code to check if a distribution is discrete or continuous.

I don't quite understand the logic of `fn`, `is_discrete`, `event_dim` in those `dependent`, `dependent_property` decorators, hence would like to defer to @neerajprad to review it. :)

**Comment by fritzo** [comment]
> ... the logic of fn, is_discrete, event_dim in those dependent, dependent_property decorators

To clarify, I have added `.__call__()` methods to `dependent` and `dependent_property`. That is because I would like to support overriding of default `.is_discrete` and `.event_dim` using syntax like
```py
arg_constraints = {"x": constraints.dependent(event_dim=1)}

@constraints.dependent_property(is_discrete=True)
def support(self):
    ...
```
But to be compatible with the old syntax without parens, we can't exactly add those new parameters to the constructor. So instead I've allowed creating of a temporary `dependent` constraint and then use its `.__call__()` method to create objects with the right parameters. @fehiepsi we use this same trick in Pyro, e.g. to create `PyroSample` objects using syntax either with or without parens:
```py
@PyroSample
def x():
    ...

@PyroSample(infer={...})
def x():
    ...
```

**Comment by fehiepsi** [comment]
Thanks @fritzo, that makes sense to me! Luckily we don't use `dependent`, `dependent_property` in numpyro (there it might not be easy to overwrite the `__call__` method, which has the same functionality as `.check`) except for Uniform parameters (which should have the default `event_dim`, `is_discrete` properties).

File: torch/distributions/constraints.py
Comment (neerajprad) [inline torch/distributions/constraints.py:149]: Do we need to unwrap if the `base_dist` is itself an instance of `Independent`? I'm asking because I'm not sure if `biject_to` will work without that.

File: torch/distributions/constraint_registry.py
Comment (neerajprad) [inline torch/distributions/constraint_registry.py:163]: I'm not sure if this will be enough to address #50496 since the event dim information will be lost when we call `log_abs_det_jacobian`, i.e. it seems to me that the transform returned for both the independent as well as the base constraint is the same. One way would be to post-hoc modify the output event dim of a transform to handle that use case. @feynmanliang - please correct me if I missed something.

File: torch/distributions/constraints.py
Comment (neerajprad) [inline torch/distributions/constraints.py:164]: Should we also raise an error here if the value's ndim < event_dim ? e.g. something like `independent(constraints.positive, 2).check(torch.ones(10))` should probably throw an error. WDYT?

File: torch/distributions/constraint_registry.py
Comment (fehiepsi) [reply torch/distributions/constraint_registry.py:163]: I just think that we no longer need `input_event_dim`, `output_event_dim` in transforms. All we need is to define a correct `domain`, `codomain`. For composed transform, we still need to handle the logic properly.

Comment (neerajprad) [reply torch/distributions/constraint_registry.py:163]: Interesting, yes that's one of the differences w.r.t. numpyro. That's probably worth another discussion.

File: torch/distributions/multinomial.py
Comment (fehiepsi) [inline torch/distributions/multinomial.py:79]: I think it is better to have `constraints.multinomial`.

File: torch/distributions/constraint_registry.py
Comment (neerajprad) [reply torch/distributions/constraint_registry.py:163]: @fritzo - I like @fehiepsi's suggestion of modifying the domain/codomain's event dim (and remove input/output event dims since that wasn't part of last release) to handle this, but regardless, all the proposed changes in this PR look great to me, so please feel free to defer this to later.

Comment (fritzo) [reply torch/distributions/constraint_registry.py:163]: done

File: torch/distributions/constraints.py
Comment (fritzo) [reply torch/distributions/constraints.py:151]: done

File: torch/distributions/constraints.py
Comment (fritzo) [reply torch/distributions/constraints.py:166]: done

File: torch/distributions/multinomial.py
Comment (fritzo) [reply torch/distributions/multinomial.py]: Good idea. I've added a `constraints.multinomial` in this PR. However note that it checks the slightly weaker condition
```diff
- (0 <= x).all(-1) & (x.sum(-1) == total_count)
+ (0 <= x).all(-1) & (x.sum(-1) <= total_count)
```
because we allow passing a bogus `total_count.max()` to `Multinomial` when it is used as a likelihood. I plan to try to address this in the future while fixing #42407.

File: torch/distributions/constraint_registry.py
Comment (fritzo) [reply torch/distributions/constraint_registry.py:163]: That sounds right. I think we can follow @feynmanliang's suggestion in a future PR and say wrap with an `IndependentTransform` or set the `Transform.event_dim`. I'll demote this PR from "Fixes" to "Addresses".

**Review by neerajprad — APPROVED** [review]

**Comment by neerajprad** [comment]
There are some internal test failures that I'm blocked on and still looking into.

File: torch/distributions/lowrank_multivariate_normal.py
Comment (neerajprad) [inline torch/distributions/lowrank_multivariate_normal.py:79]: Note that the internal failures are due to checks like `if dist.support is constraints.real` or `isinstance(dist.support, constraints._Real)`. For some of the distributions, this check will need to be updated (e.g. multinomial and I don't see a way around that) in client code. Constraints wrapped within `Independent` will be harder to check this way. e.g. if we have `independent(real, 2)`, what would be the recommended way to check the base constraint without the `event dim` information?

Comment (neerajprad) [reply torch/distributions/lowrank_multivariate_normal.py:79]: This points to a more general issue that I have observed, which is what is the recommended way for inferring constraint type. e.g. the most general way would be something like `isinstance(constraint, _IntegerInterval)` but that requires peeking into the non-public interface.

Comment (fritzo) [reply torch/distributions/lowrank_multivariate_normal.py:79]: @neerajprad for `constraints.independent(constraints.real, 2)` I suppose we could brute force and
```py
assert isinstance(c, constraints.independent)
assert isinstance(c.base_constraint, constraints.real)
assert isinstance(c.event_dim, 2)
```
or we could define `.__eq__()` to enable syntax like:
```py
assert c == constraints.independent(constraints.real, 2)
```
(we already do this for transforms).

Comment (neerajprad) [reply torch/distributions/lowrank_multivariate_normal.py:79]: There are, I think, two issues - I created https://github.com/pytorch/pytorch/issues/50616 to discuss one of these. The other issue is checking for the constraint type (disregarding event dim and the wrapping `independent` constraint) which is where most of the usage is rather than a hard equality check. So something like, `isinstance(constraint, constraints._Real) or (isinstance(constraint, constraints._Independent) and isinstance(constraint.base_constraint, constraints._Real))` will work but only if the `independent`s aren't nested.

Comment (neerajprad) [reply torch/distributions/lowrank_multivariate_normal.py:79]: I plan to use something like the following internally to handle these use cases, and we can consider providing this as a utility in PyTorch itself so as to avoid usage of non-public classes for these kind of checks. I am not sure how much usage like this we are going to find in the wild, so this may still end up breaking some code, which is my only (although minor) concern.

```python
def _unwrap(constraint):
    if isinstance(constraint, constraints.independent):
        return _unwrap(constraint.base_constraint)
    return constraint if isinstance(constraint, type) else constraint.__class__


def constraint_type_eq(constraint1, constraint2):
    return _unwrap(constraint1) == _unwrap(constraint2)
```

**Pull request closed without merging**
